Fix: Do not merge union types in truthy filter

straight-shoota opened the pull request:
Resolves #12751

asterite commented on spec/compiler/semantic/if_spec.cr:
```diff
@@ -235,6 +235,34 @@ describe "Semantic: if" do
+      end
+
+      a = Bar.new.as(Foo)
+      if b.as?(M)
```
```suggestion
      if a.as?(M)
```

asterite commented on spec/compiler/semantic/if_spec.cr:
```diff
@@ -235,6 +235,34 @@ describe "Semantic: if" do
+
+      a = Bar.new.as(Foo)
+      if b.as?(M)
+        b.m
```
```suggestion
        a.m
```

HertzDevil commented:
`as?` alone has never acted as a type filter:

```crystal
x = 1 || ""
if x.as?(String)
  x.codepoints # Error: undefined method 'codepoints' for Int32 (compile-time type is (Int32 | String))
end
```

Assignment to a separate variable is necessary here.

straight-shoota commented:
Yes, that's what the spec is supposed to do. Seems to have ended up differently when I refactored it before publishing the PR, and the edit suggestions didn't help either 🙈 

Correct example is in https://github.com/crystal-lang/crystal/issues/12751

Fixed it now.
